ecs-init/engine/engine.go

```diff
@@ -180,15 +189,23 @@ func (e *Engine) StartSupervised() error {
 			return engineError("could not start Agent", err)
 		}
 		log.Infof("Agent exited with code %d", agentExitCode)
-		if agentExitCode == upgradeAgentExitCode {
+
+		switch agentExitCode {
+		case upgradeAgentExitCode:
 			err = e.upgradeAgent()
 			if err != nil {
 				log.Error("could not upgrade agent", err)
+			} else {
+				continue
```

Comment:
I had to read through a few times to confirm that the intent with this `continue` is to intentionally start the Agent again and skip the retry given an upgraded image. Can we drop a comment here about restarting with the successfully upgraded agent? Is there reason to not try to start the agent as quickly as possible if it didn't upgrade successfully? Perhaps the upgrade handling logic could be simplified and handle this:

```go
if err != nil {
  log.Errorf("failed to upgrade agent: %s", err)
}
// start up the image loaded - upgraded or existing.
continue
```

---

It's very out of scope here but is however worth the thought exercise: the intent and handling would be clearer if the code was reasonable to tail-call itself in order to "start" again. ex:

```golang
upgradeErr := e.upgradeAgent()
if upgradeErr != nil {
  ... // log, handle "rollback" / stay-in-place start
}
return e.StartSupervised()
```

This is not going to happen here, but these handlers should be lower level concerns than what's going on with the supervision process.

Reply:
sure I will leave a comment

Reply:
Yeah, I meant more that commentary for `continue` is surrounding its intent, which is to go straight to starting the upgraded Agent. The `if {} else {}` can go away either way.